### PR TITLE
Add Request() method to structs

### PR DESCRIPTION
**Pull request by itsjamie**
In addition, moved the structs to seperate files to closely match the
layout as defined in the RFC23 in the table of contents where it lists
the various tags.

### CODE REVIEW

File: hls/decode-util_test.go [hunk 1]
```diff
@@ -250,8 +249,8 @@ func TestReadMediaPlaylist(t *testing.T) {
 		if s.URI != newP.Segments[i].URI {
 			t.Errorf("Expected URI to be %s, but got %s", s.URI, newP.Segments[i].URI)
 		}
-		if s.Map != nil && !reflect.DeepEqual(s.Map, newP.Segments[i].Map) {
-			t.Errorf("Expected %d Segment Map tp be %v, but got %v", i, s.Map, newP.Segments[i].Map)
+		if s.Map != nil && s.Map.URI != newP.Segments[i].Map.URI {
```
Comment (itsjamie) [inline hls/decode-util_test.go]: The private reference to the playlist was breaking this test.

File: hls/decode.go [hunk 2]
```diff
@@ -19,7 +19,9 @@ func (p *MasterPlaylist) Parse(reader io.Reader) error {
 
 	// Parsed Data
 	var renditions []*Rendition
-	variant := &Variant{}
+	variant := &Variant{
```
Comment (itsjamie) [inline hls/decode.go:22]: When making these changes, I found that the decode methods below actually created new structs, we could make those decode methods part of the struct, maybe implemented as `UnmarshalText(text []byte) error` to implement `encoding.TextUnmarshaller` interface?

Comment (itsjamie) [reply hls/decode.go:22]: In addition, since they operate on a pointer, they could use these structs.

File: hls/decode-util_test.go (hunk 1 above)
Comment (bradleyfalzon) [reply hls/decode-util_test.go]: Yeah, this is a pita, check out https://github.com/google/go-cmp if you haven't seen it already.

Specifically:
```
equality is determined by recursively comparing the primitive kinds on both values,
much like reflect.DeepEqual. Unlike reflect.DeepEqual, unexported fields are not 
compared; they result in panics unless suppressed by using an Ignore option.
```

I understand the desire to not include external dependencies, so it's just a suggestion.

**Review by mikerudolph — APPROVED** [review]
LGTM

**Review by tainacleal — APPROVED** [review]

File: hls/decode-util_test.go (hunk 1 above)
Comment (itsjamie) [reply hls/decode-util_test.go]: Definitely, thanks for that go-cmp link @bradleyfalzon . Led me to the idea to just create a Equal() function for the type. If we ever use the go-cmp for the check it'll use it as well.

Appreciate it!

File: hls/decode-util_test.go [hunk 3]
```diff
@@ -250,8 +250,8 @@ func TestReadMediaPlaylist(t *testing.T) {
 		if s.URI != newP.Segments[i].URI {
 			t.Errorf("Expected URI to be %s, but got %s", s.URI, newP.Segments[i].URI)
 		}
-		if s.Map != nil && !reflect.DeepEqual(s.Map, newP.Segments[i].Map) {
-			t.Errorf("Expected %d Segment Map tp be %v, but got %v", i, s.Map, newP.Segments[i].Map)
+		if !s.Map.Equal(newP.Segments[i].Map) {
```
Comment (tainacleal) [inline hls/decode-util_test.go:253]: Joe Tsai covered this on his talk on gophercon!
https://about.sourcegraph.com/go/forward-compatible-go-code

I think it was the same day you opened this PR heh

**Pull request merged**